beacon-chain/utils/shuffle.go

```diff
@@ -70,3 +79,80 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
+//     """
+//     Return `p(index)` in a pseudorandom permutation `p` of `0...list_size - 1` with ``seed`` as entropy.
+//     Utilizes 'swap or not' shuffling found in
+//     https://link.springer.com/content/pdf/10.1007%2F978-3-642-32009-5_1.pdf
+//     See the 'generalized domain' algorithm on page 3.
+//     """
+//     assert index < list_size
+//     assert list_size <= 2**40
+//     for round in range(SHUFFLE_ROUND_COUNT):
+//         pivot = bytes_to_int(hash(seed + int_to_bytes1(round))[0:8]) % list_size
+//         flip = (pivot - index) % list_size
+//         position = max(index, flip)
+//         source = hash(seed + int_to_bytes1(round) + int_to_bytes4(position // 256))
+//         byte = source[(position % 256) // 8]
+//         bit = (byte >> (position % 8)) % 2
+//         index = flip if bit else index
+//     return index
+func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error) {
+	if params.BeaconConfig().ShuffleRoundCount == 0 {
+		return index, nil
+	}
+	// spec: assert index < list_size
+	if index >= listSize {
+		return 0, fmt.Errorf("input index %d out of bounds: %d",
+			index, listSize)
+	}
+	// spec: assert list_size <= 2**40
+	if listSize > uint64(math.Pow(2, 40)) {
+		return 0, fmt.Errorf("list size %d out of bounds",
+			listSize)
+	}
+	buf := make([]byte, hTotalSize, hTotalSize)
+	// Seed is always the first 32 bytes of the hash input, we never have to change this part of the buffer.
+	copy(buf[:32], seed[:])
+	for round := uint8(0); round < uint8(params.BeaconConfig().ShuffleRoundCount); round++ {
+		buf[hSeedSize] = round
+		hash := hashutil.Hash(buf[:hPivotViewSize])
+		hash8 := hash[:8]
+		pivot := bytesutil.FromBytes8(hash8) % listSize
+		flip := (pivot - index) % listSize
+		// spec: position = max(index, flip)
+		// Why? Don't do double work: we consider every pair only once.
+		// (Otherwise we would swap it back in place)
+		// Pick the highest index of the pair as position to retrieve randomness with.
+		position := index
+		if flip > position {
+			position = flip
+		}
+		// spec: source = hash(seed + int_to_bytes1(round) + int_to_bytes4(position // 256))
+		// - seed is still in 0:32 (excl., 32 bytes)
+		// - round number is still in 32
+		// - mix in the position for randomness, except the last byte of it,
+		//     which will be used later to select a bit from the resulting hash.
+		p4b := bytesutil.Bytes4(position >> 8)
+		copy(buf[hPivotViewSize:], p4b[:])
+		source := hashutil.Hash(buf)
+		// spec: byte = source[(position % 256) // 8]
+		// Effectively keep the first 5 bits of the byte value of the position,
+		//  and use it to retrieve one of the 32 (= 2^5) bytes of the hash.
+		byteV := source[(position&0xff)>>3]
+		// Using the last 3 bits of the position-byte, determine which bit to get from the hash-byte (8 bits, = 2^3)
+		// spec: bit = (byte >> (position % 8)) % 2
+		bitV := (byteV >> (position & 0x7)) & 0x1
+		//index = flip if bit else index
```

Comment:
??

Reply:
it implements the spec in a little different way. practically doing the same without math operations